### PR TITLE
216 visualize cost and emissions operation

### PR DESCRIPTION
**Pull request by daren-thomas**
This PR implements #216 - it shows three plots inside a new "title bar" region of the visualizer. Not everything is defined in the spec, so let's see how close I got and then fix the remaining bits.

There's an atavistic arrow that will become the dropdown for choosing the main plot in the future.

image

Note that it's also possible to click on those three numbers: That will turn it from total to specific (/m2)

image

The data is dummy data - but is passed in to the plots as a `delegate` that takes a `results` object - @christophwaibel as soon as you know how to compute the values, I can plug it right in.

@amraladdin check this out to see if it's what you wanted.

### CODE REVIEW

**Comment by amraladdin** [comment]
Looks great Daren!! The position of the 3 values, the arrow top left. I also think this font would work just fine for the MVP. We might not even have to worry about a new font till end of the project lifetime.

I would adjust one point before making the pull request if possible. It's quite important.
To move the Legend to a horizontal position rather than vertical, in a way that wouldn't allow it to overlap with the data bars. See below:
200701_Demand_Monthly_Edit

There's another, less important, but if it is easy for you to change, I would do it right away. It's about the 3 plots top right. I would change to have the units under the value. See below:
image

**Comment by daren-thomas** [comment]
@amraladdin  the legend of the plot is not part of this issue and I won't be changing it in this PR.

The colours really do seem off. I copied them from the specification in the issue.

> One last thing... The RGB values of the left box seem off... I might have given you the wrong values. The colour should be yellow-ish (see the 2nd image I added in this comment). The colour name is: e1f21f and The RGB values should be:
225, 242, 31

This comment doesn't really help at all because there are _three_ colours - which one are you referring to?

ugh! having the units below.... do you know how much time I spent getting them above? but sure. I'll change that. (this is why having a good mockup is important - it's easier to do these changes in illustrator than to do program them and then iterate...)

also, you never mentioned the background colour... that means you're ok with it being grey? I think it looks really bad.

I also suggest the down arrow be much bigger - in fact, the same dimensions as the boxes to the right...

**Comment by christophwaibel** [comment]
upvote for white background

**Comment by daren-thomas** [comment]
image

@amraladdin this is how it looks right now. i removed the left and right arrows, made the dropdown arrow larger (and hooked up the right-arrow function to it as the dropdown list is a separate issue).

i'll be working on cea tomorrow, so if you and @christophwaibel could check this and possibly merge, i could continue with issue #309 (dropdown list of plots)

**Comment by amraladdin** [comment]
Hey @daren-thomas ,

> AE: One last thing... The RGB values of the left box seem off... I might have given you the wrong values. The colour should be yellow-ish (see the 2nd image I added in this comment). The colour name is: e1f21f and The RGB values should be:
225, 242, 31
> DT: This comment doesn't really help at all because there are _three_ colours - which one are you referring to?
You're absolutely right. my bad. I'm talking about the energy (kWh) color. The one on the left. However, I see you figured that out on your own. Good job!

> ugh! having the units below.... do you know how much time I spent getting them above? but sure. I'll change that. (this is why having a good mockup is important - it's easier to do these changes in illustrator than to do program them and then iterate...)
I agree. Having a good mockup is important, but given the time and resources, this is the best you'll get for now :)
I don't mind learning illustrator or Adobe XD (better for prototyping) if that's what it'll take. I'll need 3-5 days to do so though.

> also, you never mentioned the background colour... that means you're ok with it being grey? I think it looks really bad.
I'm not sure to which level of detail you'll need the graphs. Think that I'm also tight on time, so I try to choose the important aspects of a Graph and leave other decisions to you to do, thinking that we will iterate on it later on. If that's bad practice, we need to discuss this in a team meeting.
Besides, we had a full Hivathon day sitting together, and a week before when I submitted the graphs. Just write me what's missing on the issue and I'll get back to you with an answer right away :)

> I also suggest the down arrow be much bigger - in fact, the same dimensions as the boxes to the right...
Definitely not :D I'm sorry man, but before it was much better. Now it's too big, you can't see the bar chart :D

File: Mothercell/Hive.IO/Hive.IO/Building/Zone.cs [hunk 1]
```diff
@@ -51,7 +51,7 @@ public class Zone
         /// <summary>
         /// Internal loads.
         /// Values indicate full / maximal value.
-        /// Unit in W/sqm
+        /// UnitText in W/sqm
```
Comment (christophwaibel) [inline Mothercell/Hive.IO/Hive.IO/Building/Zone.cs:54]: why change to `UnitText`?

File: Mothercell/Hive.IO/Hive.IO/DataHandling/Results.cs [hunk 2]
```diff
@@ -72,7 +72,7 @@ public class Results
         /// </summary>
         public bool[,] SupplyTypes { get; private set; }
         /// <summary>
-        /// Capacity per technology. Unit is defined in 'SupplyCapUnits
+        /// Capacity per technology. UnitText is defined in 'SupplyCapUnits
```
Comment (christophwaibel) [inline Mothercell/Hive.IO/Hive.IO/DataHandling/Results.cs:75]: same as previous comment

File: Mothercell/Hive.IO/Hive.IO/Building/Zone.cs (hunk 1 above)
Comment (christophwaibel) [inline Mothercell/Hive.IO/Hive.IO/Building/Zone.cs:54]: why change to UnitText?

File: Mothercell/Hive.IO/Hive.IO/Plots/OperationalPerformancePlot.cs [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+                graphics.FillRectangle(new SolidBrush(_properties.Color), box);
+            }
+
+            // center data in the box
+            var data = Data(results);
+            var dataSize = GH_FontServer.MeasureString(data, _boldFont);
+            var dataX = bounds.Left + (bounds.Width - dataSize.Width) / 2;
+            var dataY = bounds.Top + bounds.Height / 2 - (float)dataSize.Height / 2;
+            graphics.DrawString(data, _boldFont, brush, dataX, dataY);
+            RhinoApp.WriteLine($"Rendering {data} to {dataX}, {dataY}");
```
Comment (christophwaibel) [inline Mothercell/Hive.IO/Hive.IO/Plots/OperationalPerformancePlot.cs:89]: `RhinoApp.WriteLine($"Rendering {data} to {dataX}, {dataY}");`

I would deactivate this. It spams the command line in Rhino every time you move the Visualizer component or change its size etc. Usually, you wanna keep the rhino command line clean so you can track the actual Rhino commands that you've used for drawing geometry

File: Mothercell/Hive.IO/Hive.IO/Plots/OperationalPerformancePlot.cs [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+            // center unit below data
+            var unitX = bounds.Left + (bounds.Width - GH_FontServer.StringWidth(UnitText, _standardFont)) / 2;
+            var unitY = dataY + dataSize.Height;
+            graphics.DrawString(UnitText, _standardFont, brush, unitX, unitY);
+            RhinoApp.WriteLine($"Rendering {UnitText} to {unitX}, {unitY}");
```
Comment (christophwaibel) [inline Mothercell/Hive.IO/Hive.IO/Plots/OperationalPerformancePlot.cs:95]: same, I'd deactivate the `WriteLine`

File: Mothercell/Hive.IO/Hive.IO/Plots/OperationalPerformancePlot.cs [hunk 5]
```diff
@@ -0,0 +1,103 @@
+﻿using System;
+using System.Drawing;
+using Grasshopper.GUI.Canvas;
+using Grasshopper.Kernel;
+using Rhino;
+
+namespace Hive.IO.Plots
+{
+    public delegate double QueryResults(Results results);
```
Comment (christophwaibel) [inline Mothercell/Hive.IO/Hive.IO/Plots/OperationalPerformancePlot.cs:9]: note for your delegate that `Results` might contain other objects (`Hive.IO.EnergySystems`, `Hive.IO.Building`, etc... not final yet, but that's the concept), and not properties of type `double` anymore. So e.g. if you wanna get heating loads, you will need to read something like `Results.Building.Zone[i].HeatingLoads` . 

I can make getters in the `Results` class tho, so the API doesn't become so nested!

File: Mothercell/Hive.IO/Hive.IO/DataHandling/Results.cs (hunk 2 above)
Comment (daren-thomas) [reply Mothercell/Hive.IO/Hive.IO/DataHandling/Results.cs:75]: ugh! that was a refactoring gone wrong... I'll fix it.

File: Mothercell/Hive.IO/Hive.IO/Building/Zone.cs (hunk 1 above)
Comment (daren-thomas) [reply Mothercell/Hive.IO/Hive.IO/Building/Zone.cs:54]: at least it's only doing it in _comments_...

File: Mothercell/Hive.IO/Hive.IO/Plots/OperationalPerformancePlot.cs (hunk 3 above)
Comment (daren-thomas) [reply Mothercell/Hive.IO/Hive.IO/Plots/OperationalPerformancePlot.cs:89]: good point. I'll delete all the debug output - was just using that for creating the tool...

File: Mothercell/Hive.IO/Hive.IO/Plots/OperationalPerformancePlot.cs (hunk 4 above)
Comment (daren-thomas) [reply Mothercell/Hive.IO/Hive.IO/Plots/OperationalPerformancePlot.cs:95]: ditto

File: Mothercell/Hive.IO/Hive.IO/Plots/OperationalPerformancePlot.cs (hunk 5 above)
Comment (daren-thomas) [reply Mothercell/Hive.IO/Hive.IO/Plots/OperationalPerformancePlot.cs:9]: The delegates still need to be written. using the delegate system, the visualizer is decoupled from the `Results` object, in that the delegate will define _how_ to actually query the results. So have fun! And when you're ready, we can plug it in!

File: Mothercell/Hive.IO/Hive.IO/DataHandling/Results.cs (hunk 2 above)
Comment (christophwaibel) [reply Mothercell/Hive.IO/Hive.IO/DataHandling/Results.cs]: I reversed it

File: Mothercell/Hive.IO/Hive.IO/Plots/OperationalPerformancePlot.cs (hunk 3 above)
Comment (christophwaibel) [reply Mothercell/Hive.IO/Hive.IO/Plots/OperationalPerformancePlot.cs:89]: ok cool. will you merge this branch then when you're done? cheers

**Pull request merged**